### PR TITLE
concurrency: rollback ignored sequence numbers on lock acquisition

### LINKED ISSUE #102269: concurrency: rollback locks sequence numbers on lock reacquisition for unreplicated locks
**Background**

All operations performed by a transaction in CRDB have an associated sequence number, which are used to provide relative ordering for all operations. Savepoints are built on top of this concept -- whenever a savepoint is rolled back, associated sequence numbers are considered to be "ignored". 

As of the time of writing, CRDB deviates from lock rollback compared to Postgres -- savepoint rollback does not result in lock rollbacks. This is because the client never issues an RPC to update any locks acquired under savepoints. However, the `OnLockUpdate` method on the `LockManager`, and the `concurrency_control` package in general, is designed to handle lock rollbacks. 

**Propopsal**

Locks in the lock table store all sequence numbers at which they have been acquired. Calls to `OnLockUpdate` are passed a list of `IgnoredSequenceNumbers`. These are then used to prune the list of sequence numbers (`holder.seqs`) the lock is held at; if this list becomes empty, the lock is no longer held, and removed from the lock table. 

In addition to the `OnLockUpdate` path, we should do this pruning on the lock (re-)acquisition path as well. This would entail teaching the arguments of `OnLockAcquired` (`roachpb.LockAcquisition`) to have access to a list of `IgnoredSequenceNumbers`. We could then make use of this list like we do in `OnLockUpdate`.   

Jira issue: CRDB-27375

### PR DESCRIPTION
**Pull request by arulajmani**
This patch adds logic to prune the list of sequence numbers tracked by
the lock table for unreplicated locks. This is done when some of the
tracked sequence numbers are considered ignored, by virtue of a
savepoint rollback.

Note that we only do so for unreplicated locks, and only if an
unreplicated lock is being reacquired. This is because the in-memory
lock table is only the source of truth for in-memory locks; the mvcc
keyspace is the source of truth for replicated ones. As such, trying
to mimic the logic is hard/error-prone -- so we don't attempt to do
so.

Fixes https://github.com/cockroachdb/cockroach/issues/102269

Release note: None

### CODE REVIEW

**Review by nvb — APPROVED** [review]
:lgtm: once the first commit is merged.

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103493#-:-NVjoOfY2xUrIeIgYxCA:bln91dy)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Comment by arulajmani** [comment]
TFTR! 

bors r=nvanbenschoten

**Pull request merged**
